packages/SwingSet/src/limits/calculate.js

```diff
@@ -0,0 +1,193 @@
+// @ts-check
+
+/** @typedef {import('./types').Budget} Budget */
+/** @typedef {import('./types').MemoryCostModel} MemoryCostModel */
+
+const { details: X } = assert;
+// @ts-expect-error verbatim is not defined
+const v = assert.verbatim || assert.quote;
+
+export const BIGINT_WORD_VALUE = 256n ** 8n;
+
+/**
+ * @param {MemoryCostModel} costModel
+ */
+export const makeCalculateCost = costModel => {
+  assert.typeof(
+    costModel,
+    'object',
+    X`costModel ${costModel} must be an object`,
+  );
+  const {
+    baseCost,
+    bigintPerWordCost,
+    stringPerCharacterCost,
+    objectPerPropertyCost,
+  } = costModel;
+  assert.typeof(baseCost, 'bigint', X`baseCost ${baseCost} must be a bigint`);
+  assert.typeof(
+    bigintPerWordCost,
+    'bigint',
+    X`bigintPerWordCost ${bigintPerWordCost} must be a bigint`,
+  );
+  assert.typeof(
+    stringPerCharacterCost,
+    'bigint',
+    X`stringPerCharacterCost ${stringPerCharacterCost} must be a bigint`,
+  );
+  assert.typeof(
+    objectPerPropertyCost,
+    'bigint',
+    X`objectPerPropertyCost ${objectPerPropertyCost} must be a bigint`,
+  );
+
+  let totalCost = 0n;
+  const getTotalCost = () => totalCost;
+
+  /**
+   * @param {unknown} value
+   */
+  const shallowCost = value => {
+    let cost = baseCost;
+    if (Object(value) === value) {
```

Comment:
This technique for testing object-ness allocates. We avoid it in at least one other case.

Reply:
Thanks, that's good to know!

Reply:
I'll use `@endo/marshal`'s `isObject`, and punt to their implementation for any future optimisations.